docs/proposals/20210716-async-azure-resource-creation-deletion.md

```diff
@@ -0,0 +1,224 @@
+}
+
+// AzureMachinePoolStatus defines the observed state of AzureMachinePool
+AzureMachinePoolStatus struct {
+    /*
+      Other fields omitted for brevity    
+    */
+    
+    // LongRunningOperationState saves the state for an Azure long-running operations so it can be continued on the
+    // next reconciliation loop.
+    // +optional
+    LongRunningOperationState *infrav1.Future `json:"longRunningOperationState,omitempty"`
+}
+```
+
+#### <a name='ProposedAPIChanges'></a>Proposed API Changes
+
+The proposed changes below show the changes to AzureCluster, AzureMachine, AzureMachinePool, and AzureMachinePoolMachine. The existing `LongRunningOperationState` field in AzureMachinePoolStatus will be pluralized to `LongRunningOperationStates` to store a list of Futures, following a similar pattern than Conditions, and will be extended to other CAPZ CRDs. Each Future has an identifying `Name` that ties it to a specific resource.
+
+```go
+type Futures []Future
+
+// AzureClusterStatus defines the observed state of AzureCluster.
+type AzureClusterStatus struct {
+    /*
+      Other fields omitted for brevity    
+    */
+
+    // LongRunningOperationStates saves the states for Azure long-running operations so they can be continued on the
+    // next reconciliation loop.
+    // +optional
+    LongRunningOperationStates Futures `json:"longRunningOperationState,omitempty"`
+}
+
+// AzureMachineStatus defines the observed state of AzureMachine.
+type AzureMachineStatus struct {
+    /*
+      Other fields omitted for brevity    
+    */
+
+    // LongRunningOperationStates saves the states for Azure long-running operations so they can be continued on the
+    // next reconciliation loop.
+    // +optional
+    LongRunningOperationStates Futures `json:"longRunningOperationState,omitempty"`
+}
+
+// AzureMachinePool defines the observed state of AzureMachinePool.
+type AzureMachinePool struct {
+    /*
+      Other fields omitted for brevity    
+    */
+
+    // LongRunningOperationStates saves the states for Azure long-running operations so they can be continued on the
+    // next reconciliation loop.
+    // +optional
+    LongRunningOperationStates Futures `json:"longRunningOperationState,omitempty"`
+}
+
+// AzureMachinePoolMachineStatus defines the observed state of AzureMachinePoolMachine.
+type AzureMachinePoolMachine struct {
+    /*
+      Other fields omitted for brevity
+    */
+
+    // LongRunningOperationStates saves the states for Azure long-running operations so they can be continued on the
+    // next reconciliation loop.
+    // +optional
+    LongRunningOperationStates Futures `json:"longRunningOperationState,omitempty"`
+}
+
+```
+
+#### <a name='ProposedControllerChanges'></a>Proposed Controller Changes
+
+* Add an `AzureClientTimeout` which is the duration after which an Azure operation is considered a long running operation which should be handled asynchronously. Proposed starting value is 5 seconds.
```

Comment:
Would this timeout be for each call to Azure, the entire reconciliation loop, or a smaller scope, like all of the calls to Azure excluding other activities in the reconciliation loop? The reason I ask is that if the reconciliation loop contains multiple serial Azure requests, it's conceivable that the whole loop could take much longer than 5 seconds. Say there are 4 resources which take 5 seconds a piece, 5s + 5s + 5s + 5s = 20s, as a worse case it would not be very responsive. Also, a long running reconciliation loop including activities other than just Azure requests might not complete in 5s and end up in a state where it can no longer make progress. Thoughts?

Reply:
that's a good question; my original thinking is that the 5s timeout, which is quite aggressive, would be per operation (each call). Then we'd have another overall reconcile loop timeout. Currently, that timeout is set at 90 minutes (which is also the ARM timeout) but I think with this proposal we could bring that down to something more responsive, like 30s.

The specific numbers are not set in stone, I need to do some performance testing with different values and calculate the P99 expected completion time of operations that are not long-running. 

The other question is whether we should have the same timeout value for all operations (the 5s) or curate per operation. For simplicity, I think we should start with a single value. Later on, we might want to optimize by calculating a dynamic timeout value for each operation based on heuristics. That would be better IMO than statically defining artificial timeout durations for each operation which might vary over time and might not be the same depending on region, subscription, etc.

Reply:
Added some more details about this in the Context Timeouts section of proposed changes, and an open question